src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs

```diff
@@ -31,12 +31,9 @@ public Task<ProjectStateChecksums> GetStateChecksumsAsync(CancellationToken canc
         {
             using (Logger.LogBlock(FunctionId.ProjectState_ComputeChecksumsAsync, FilePath, cancellationToken))
             {
-                // get states by id order to have deterministic checksum
-                var orderedDocumentIds = ChecksumCache.GetOrCreate(DocumentIds, _ => DocumentIds.OrderBy(id => id.Id).ToImmutableArray());
-                var documentChecksumsTasks = orderedDocumentIds.Select(id => DocumentStates[id].GetChecksumAsync(cancellationToken));
-
-                var orderedAdditionalDocumentIds = ChecksumCache.GetOrCreate(AdditionalDocumentIds, _ => AdditionalDocumentIds.OrderBy(id => id.Id).ToImmutableArray());
-                var additionalDocumentChecksumTasks = orderedAdditionalDocumentIds.Select(id => AdditionalDocumentStates[id].GetChecksumAsync(cancellationToken));
+                // get states by document order to have deterministic checksum
```

Comment:
This loses the `OrderBy` that used to be there.

Reply:
we need the orderby until this is done (https://github.com/dotnet/roslyn/issues/22496)

Reply:
@heejaechang How much work is that then? I can leave the ID ordering, but the memory cost will then be non-trivial in the cases that @sharwell identified. If it's something I can just fix I'd like to do so.

Reply:
I need to add WithDocumentIds and WithAdditionalDocumentIds and WithProjectIds to Project/ProjectState and Solution/SolutionState

and give projectId order as it got and documentid order as it to to it from here and here
http://source.roslyn.io/#Microsoft.CodeAnalysis.Remote.Workspaces/Services/SolutionCreator.cs,102

http://source.roslyn.io/#Microsoft.CodeAnalysis.Remote.Workspaces/Services/SolutionCreator.cs,283

Reply:
but ordering happens not that frequently though. and ordering can also use ImmutableList isn't it if one big chunk of memory (array) is the issue.

Reply:
OK, I see why it's necessary now. I was thinking that once we sent something to the OOP process we weren't recomputing checksums on that side _too_, since we're in theory recomputing a value we already knew.

We might have a better fix available because the ProjectState already has a ImmutableDictionary that it uses to store the data to the DocumentStates. [This](https://msdn.microsoft.com/en-us/magazine/mt795189.aspx) implies that the underlying structure for that ImmutableDictionary is already a tree -- we could just switch that to ImmutableSortedDictionary (already a tree), and then we already have the keys in sorted order there. I'm not an expert in the subtleties of the data structures here, but it might be equal cost and then we just avoid all the sorting. If we have an expert that can confirm that the performance of ImmutableDictionary and ImmutableSortedDictionary would be the same (or nearly so) in this case since we can apply the same ordering you're doing with the GUID, then let's just do that.

Just moving to ImmutableList worries me because if there's a large number of adds coming in, you're still resorting from scratch each time. Although we'd avoid the large allocations, the total allocation size would actually go _up_. It all depends exactly on how often this were to run during a large batch add case.

Reply:
I believe @AArnott is pretty familiar with the collections involved.

Reply:
I want, at some point, VS not send all data but let OOP to read data itself and then only a corner case where data got changed once VS read the data, it gets data from VS. after that to verify it has all same data, two must be able to verify data independently. so OOP re-calculate checksum themselves and verify it got all data correctly. even though for now, data all came from VS.

Reply:
we can do that as well. at some point, I will fix the bug (#22496) regardless.

Reply:
@heejaechang Done, so please re-review. We'll now just use ImmutableSortedDictionary (which we believe to be as fast), but we'll avoid the allocation and intermediate sorting.

Reply:
Yes, I can confirm that `ImmutableDictionary` and `ImmutableSortedDictionary` have nearly equivalent trees and algorithmic complexity. In fact, `ImmutableDictionary` internally hashes its keys and then stores them using the same tree structure as `ImmutableSortedDictionary`. So leaving the hashing out of it and just sorting the keys may be slightly faster than `ImmutableDictionary` (provided your CompareTo routine is fast).

Reply:
@AArnott In this case it's a battle between Guid.GetHashCode and Guid.CompareTo...which is probably not measurable.